server/server.go

```diff
@@ -0,0 +1,157 @@
+package main
+
+import (
+	"flag"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/allegro/bigcache"
+)
+
+const (
+	apiVersion  = "v1"
+	apiBasePath = "/api/" + apiVersion + "/"
+
+	// path to cache.
+	cachePath = apiBasePath + "cache/"
+)
+
+var (
+	lifeWindow int64
+	hardMax    int
+	verbose    bool
+	logfile    string
+	port       int
+	cache      *bigcache.BigCache
+)
+
+func init() {
+	flag.BoolVar(&verbose, "v", false, "Verbose logging.")
+	flag.Int64Var(&lifeWindow, "lifetime", 10, "Lifetime, in minutes, of each cache object.")
+	flag.IntVar(&hardMax, "max", 8192, "Maximum amount of objects in the cache.")
+	flag.IntVar(&port, "port", 9090, "The port to listen on.")
+	flag.StringVar(&logfile, "logfile", "", "Location of the logfile.")
+}
+
+func main() {
+	flag.Parse()
+
+	var logger *log.Logger
+
+	if logfile == "" {
+		logger = log.New(os.Stdout, "", log.LstdFlags)
+	} else {
+		f, err := os.OpenFile(logfile, os.O_APPEND|os.O_WRONLY, 0600)
+		if err != nil {
+			panic(err)
+		}
+		logger = log.New(f, "", log.LstdFlags)
+	}
+
+	config := bigcache.Config{
+		Shards:             1024,
+		LifeWindow:         time.Duration(lifeWindow) * time.Minute,
+		MaxEntriesInWindow: 1000 * 10 * 60,
+		MaxEntrySize:       500,
+		Verbose:            verbose,
+		HardMaxCacheSize:   hardMax,
+		OnRemove:           nil,
+		Logger:             logger,
+	}
+
+	var err error
+	cache, err = bigcache.NewBigCache(config)
+	if err != nil {
+		logger.Fatal(err)
+	}
+
+	logger.Print("cache initialised.")
+
+	// let the middleware log.
+	http.Handle(cachePath, serviceLoader(cacheIndexHandler(), requestMetrics(logger)))
+
+	logger.Printf("starting server on :%d", port)
+
+	strPort := ":" + strconv.Itoa(port)
+	log.Fatal("ListenAndServe: ", http.ListenAndServe(strPort, nil))
+}
+
+// our base middleware implementation.
+type service func(http.Handler) http.Handler
+
+// chain load middleware services.
+func serviceLoader(h http.Handler, svcs ...service) http.Handler {
+	for _, svc := range svcs {
+		h = svc(h)
+	}
+	return h
+}
+
+func cacheIndexHandler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.Method {
+		case http.MethodGet:
+			getCacheHandler(w, r)
+		case http.MethodPut:
+			putCacheHandler(w, r)
+		}
+	})
+}
+
+// middleware for request length metrics.
+func requestMetrics(l *log.Logger) service {
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			start := time.Now()
+			h.ServeHTTP(w, r)
+			l.Printf("request took %vns.", time.Now().Sub(start).Nanoseconds())
+		})
+	}
+}
+
+// handles get requests.
+func getCacheHandler(w http.ResponseWriter, r *http.Request) {
+	target := r.URL.Path[len(cachePath):]
+	if target == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte("can't get a key if there is no key."))
+		log.Print("empty request.")
+		return
+	}
+	entry, err := cache.Get(target)
```

Comment:
I tried to do a `string.Contains(err.Error(), "not found")` search on the error so I could give a 404 instead of a 500 if an item didn't exist, but I kept running into nil reference errors when trying to read the error message. If you know a way to do that, I can totally implement it as part of this.

Reply:
You may first check if `err` is not `nil` and if not then use `string.Contains`

Reply:
Turns out I had to dereference the error.